### PR TITLE
ASB-24561: Add Documentation

### PR DESCRIPTION
**Pull request by jwfraustro**
Adds a documentation section containing:

Quickstart guide
Installation
List of supported protocols and examples / output for each model
API-level model documentation

Switches in-code docstrings to Google-style.

available on https://vo-models.readthedocs.io/en/latest/

### CODE REVIEW

File: docs/source/index.rst [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+- **Pydantic-xml Models:** The project includes Python models for IVOA protocols, using `pydantic-xml <https://github.com/dapper91/pydantic-xml>`_. Based on `Pydantic <https://docs.pydantic.dev/latest/>`_, these models describe transactions for an IVOA protocol, such as UWS, and feature automatic validation, parsing and serialization of XML data for use with Python clients and web frameworks.
+
+Using the models in your project for validation and serialization to XML is as simple as:
+
+.. code-block:: python
+
+    from vo_models.uws import ShortJobDescription
+
+    job = ShortJobDescription(
+        phase="PENDING",
```
Comment (seweissman) [inline docs/source/index.rst:21]: Should the phase be set using the enum?

File: vo_models/uws/types.py [hunk 2]
```diff
@@ -7,41 +7,63 @@ class ErrorType(str, Enum):
     """Enum for error types."""
 
     TRANSIENT = "transient"
+    """The error is transient and the job may be rerun."""
     FATAL = "fatal"
+    """The error is fatal and the job may not be rerun."""
+
```
Comment (seweissman) [inline vo_models/uws/types.py:13]: What do these extra triple quoted strings in the enum do here? Are these for the docs?

**Comment by seweissman** [comment]
I checked out the RTD page and this looks great! Only have very minor comments/questions.

File: vo_models/uws/types.py (hunk 2 above)
Comment (jwfraustro) [reply vo_models/uws/types.py:13]: Yes. I've moved the rest of the projects to Google-style docstrings that are compatible with the Sphinx Napoleon autodoc extension. Using that style in Enum docstrings don't actually seem to show up in the docs, but this does. It give nice option-level doc output. 

Something like this, for example: https://vo-models.readthedocs.io/en/latest/pages/api/uws_api.html#vo_models.uws.types.ExecutionPhase

**Comment by jwfraustro** [comment]
> I checked out the RTD page and this looks great! Only have very minor comments/questions.

Thanks! I'm a big fan of the theme having a built-in dark mode.

**Review by jwfraustro — COMMENTED** [review]
I really like using the snippets in a little example file to show examples on the protocol's doc page, like this: https://vo-models.readthedocs.io/en/latest/pages/protocols/uws.html

Working on the VODataService expansion though is already demonstrating that managing separate examples is a little unwieldy. What I might end up doing is just having the examples page pull from the test files. I think this is fine for now, but I intend on probably changing that with the VODataservice PR sometime in the future.

File: vo_models/uws/types.py (hunk 2 above)
Comment (seweissman) [reply vo_models/uws/types.py:13]: Interesting. I guess I'm just not used to seeing doc strings like this.

**Review by seweissman — APPROVED** [review]
Looks good to me.

File: vo_models/uws/models.py [hunk 3]
```diff
@@ -65,15 +66,18 @@ def __init__(__pydantic_self__, **data) -> None:  # pylint: disable=no-self-argu
...
-    type (ErrorType):   Characterization of the type of the error
-    has_detail (bool):  If true then there is a more detailed error message available at /{jobs}/{job-id}/error
+    Parameters:
+        message (str):
+            (element) - A short description of the error.
+
+        type (ErrorType):
+            (attr) - Characterization of the type of the error
+        has_detail (bool):
+            (attr) - If true then there is a more detailed error message available at /{jobs}/{job-id}/error
```
Comment (falkben) [inline vo_models/uws/models.py:80]: if you haven't seen this already, you might take a look at [PEP 727](https://peps.python.org/pep-0727/), which adds documentation support to type annotations. This allows you to not have to define the type annotations twice, once as the type annotation and then again in your docstrings. It's not been accepted yet, and as far as I know, there's not a current way to get them to show up in Sphinx, but there is support for it in the [typing_extensions](https://github.com/python/typing_extensions) package and [mkdocstrings](https://github.com/mkdocstrings/mkdocstrings) for the MkDocs project. Perhaps worth an issue in the project to revisit if you don't want to deal with it now.

My understanding is that you could put the docstring helper text in as type annotations directly.

So below, for `message`, you could take out the docstring, and add to the type annotation this (maybe?):

```py
from typing import Annotated
from typing_extensions import Doc

message: Annotated[str, element, "", Doc("A short description of the error.")] = ""
```

I'm not sure I actually got the `Annotated` correct there, but there's an example in `pydantic-xml` [here](https://pydantic-xml.readthedocs.io/en/latest/pages/misc.html#optional-type-encoding)

File: vo_models/uws/models.py [hunk 4]
```diff
@@ -108,8 +113,8 @@ class ResultReference(BaseXmlModel, tag="result", ns="uws", skip_empty=True, nsm
 class Results(BaseXmlModel, tag="results", ns="uws", nsmap=NSMAP):
     """The element returned for /{jobs}/{job-id}/results
 
-    Elements:
-    result list[ResultReference]: a list of references to UWS results.
+    Parameters:
+        results (list[ResultReference]): (element) A list of references to UWS results.
```
Comment (falkben) [inline vo_models/uws/models.py:117]: this docstring doesn't include the `Optional` part to it from the type annotation

File: vo_models/uws/models.py [hunk 5]
```diff
@@ -118,20 +123,24 @@ class Results(BaseXmlModel, tag="results", ns="uws", nsmap=NSMAP):
...
+            accordance with IVOA security standards.
+        creation_time (datetime):
+            (element) - The instant at which the job was created.
+
+        job_id (str):
+            (attr) - The identifier for the job.
+        type (XlinkType):
+            (attr) - The xlink reference type of the job.
+        href (str):
+            (attr) - The link to the job.
```
[the remark this answers is not in the document]
Comment (jwfraustro) [reply vo_models/uws/models.py:143]: Damn, it looks like I missed a bunch of these. What is weird is that ones like `version` here: https://vo-models.readthedocs.io/latest/pages/api/uws_api.html#vo_models.uws.models.JobSummary.  Do correctly show as `| None` even though the docstring doesn't have it.

File: vo_models/uws/models.py [hunk 6]
```diff
@@ -167,39 +177,52 @@ class Jobs(BaseXmlModel, tag="jobs", ns="uws", nsmap=NSMAP):
...
+            (element) - The parameters to the job (where appropriate)
+        results (Results):
+            (element) - The results for the job
+        error_summary (ErrorSummary):
+            (element) - A short summary of an error
+        job_info (list[str]):
+            (element) - This is arbitrary information that can be added to the job description by the UWS
+            implementation.
+
+        version: (UWSVersion)
```
Comment (falkben) [inline vo_models/uws/models.py:221]: This is somehow being rendered as (UWSVersion | None). I'm not sure why or how.

image

File: vo_models/uws/types.py [hunk 7]
```diff
@@ -7,41 +7,63 @@ class ErrorType(str, Enum):
     """Enum for error types."""
 
     TRANSIENT = "transient"
+    """The error is transient and the job may be rerun."""
     FATAL = "fatal"
+    """The error is fatal and the job may not be rerun."""
```
Comment (falkben) [inline vo_models/uws/types.py:12]: FYI, this format was apparently proposed and rejected in `PEP 224`. I saw this briefly discussed again in PEP 727 here. https://peps.python.org/pep-0727/#string-under-definition

**Pull request merged**